gobblin-core/src/main/java/gobblin/commit/FsRenameCommitStep.java

```diff
@@ -0,0 +1,123 @@
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.commit;
+
+import java.io.IOException;
+import java.net.URI;
+
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import com.google.common.base.Optional;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.util.HadoopUtils;
+import lombok.extern.slf4j.Slf4j;
+
+
+/**
+ * A {@link CommitStep} for renaming files within a {@link FileSystem} or between {@link FileSystem}s.
+ *
+ * @author ziliu
+ */
+@Slf4j
+public class FsRenameCommitStep extends CommitStepBase {
+
+  private final Path srcPath;
+  private final Path dstPath;
+  private final String srcFsUri;
+  private final String dstFsUri;
+  private transient FileSystem srcFs;
+  private transient FileSystem dstFs;
+
+  private FsRenameCommitStep(Builder builder) throws IOException {
+    super(builder);
+    this.srcPath = builder.srcPath;
+    this.dstPath = builder.dstPath;
+    this.srcFs = builder.srcFs != null ? builder.srcFs
+        : getFileSystem(Optional.fromNullable(this.props.getProp(ConfigurationKeys.FS_URI_KEY)));
+    this.srcFsUri = this.srcFs.getUri().toString();
+    this.dstFs = builder.dstFs != null ? builder.dstFs
+        : getFileSystem(Optional.fromNullable(this.props.getProp(ConfigurationKeys.FS_URI_KEY)));
+    this.dstFsUri = this.dstFs.getUri().toString();
+  }
+
+  public static class Builder extends CommitStepBase.Builder<Builder> {
+
+    public Builder() {
+      super();
+    }
+
+    public Builder(CommitSequence.Builder commitSequenceBuilder) {
+      super(commitSequenceBuilder);
+    }
+
+    private Path srcPath;
+    private Path dstPath;
+    private FileSystem srcFs;
+    private FileSystem dstFs;
+
+    public Builder from(Path srcPath) {
+      this.srcPath = srcPath;
+      return this;
+    }
+
+    public Builder to(Path dstPath) {
+      this.dstPath = dstPath;
+      return this;
+    }
+
+    public Builder withSrcFs(FileSystem srcFs) {
+      this.srcFs = srcFs;
+      return this;
+    }
+
+    public Builder withDstFs(FileSystem dstFs) {
+      this.dstFs = dstFs;
+      return this;
+    }
+
+    @Override
+    public CommitStep build() throws IOException {
+      return new FsRenameCommitStep(this);
+    }
+  }
+
+  private FileSystem getFileSystem(Optional<String> fsUri) throws IOException {
+    if (fsUri.isPresent()) {
+      return FileSystem.get(URI.create(fsUri.get()), HadoopUtils.getConfFromState(this.props));
+    } else {
+      return FileSystem.get(HadoopUtils.getConfFromState(this.props));
+    }
+  }
+
+  @Override
+  public boolean isCompleted() throws IOException {
+    if (this.dstFs == null) {
```

Comment:
`FileSystem` cannot be serialized and deserialized by Gson. So when a `FsRenameCommitStep` is obtained from a Json String, the `FileSystem` will be null.